lib/chai/utils/eventEmitter.js

```diff
@@ -0,0 +1,89 @@
+/*!
+ * Chai - eventEmitter utility
+ * Copyright(c) 2011-2016 Jake Luer <jake@alogicalparadox.com>
+ * MIT Licensed
+ */
+
+// Global EventTarget instance
+const emitter = new EventTarget();
+
+// WeakMap: handler -> { [eventName]: listener }
+const listenerMap = new WeakMap();
+
+/**
+ * ### .on(event, handler)
+ *
+ * Adds a handler to an event.
+ *
+ *     utils.on('addProperty', function (name, getter) {
+ *         var obj = utils.flag(this, 'object');
+ *         new chai.Assertion(obj).to.be.equal(str);
+ *     });
+ *
+ * The events emitted by chai are 'addProperty' and 'addMethod'.
+ *
+ * @param {string} event to listen to
+ * @param {Function} handler function to handle the event
+ * @namespace Utils
+ * @name on
+ * @public
+ */
+export function on(event, handler) {
+  let eventMap = listenerMap.get(handler);
+  if (!eventMap) {
+    eventMap = {};
+    listenerMap.set(handler, eventMap);
+  }
+
+  if (eventMap[event])
+    return;
+
+  eventMap[event] = (e) => {
+    handler(...(e.detail || []));
+  };
+
+  emitter.addEventListener(event, eventMap[event]);
+}
```

Comment:
So it looks like this map is added to avoid the ergonomics of `e.detail`?

I think this can be removed, and instead you can subclass `Event` - as opposed to using `CustomEvent`. Something like:


```js
class PluginEvent extends Event {
  constructor(type, name, fn) {
    super(type);
    this.name = String(name)
    this.fn = fn
  }
}
```

Then simply:

```js
emitter.dispatchEvent(new PluginEvent('addMethod', name, method))
emitter.dispatchEvent(new PluginEvent('addProperty', name, getter))
```

Listeners need to change from `function(name, getter) {` to `function({name, fn}) {` but that seems reasonable to me?

Reply:
sure, just tried it out. You lose the generality of the `emit` function, but for our purposes right now that's not really a problem. Also, `emit` is largely a private function.